pair/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+                    .0,
+                amount1
+                    .overflowing_mul(total_supply)
+                    .0
+                    .overflowing_div(self.reserve1)
+                    .0,
+            )
+        }
+        if liquidity == 0 {
+            panic!("PAIR: Liquidity MUST be greater than 0.");
+        }
+        FTCore::mint(self, &to, liquidity);
+        self._update(self.balance0, self.balance1, self.reserve0, self.reserve1);
+        if fee_on {
+            self.k_last = self.reserve0.overflowing_mul(self.reserve1).0;
+        }
+        liquidity
+    }
+
+    /// Mint liquidity if fee is on.
+    /// If fee is on, mint liquidity equivalent to 1/6th of the growth in sqrt(k). So the math if the following.
+    /// Calculate the sqrt of current k using the reserves. Compare it.
+    /// If the current one is greater, than calculate the liquidity using the following formula:
+    /// liquidity = (total_supply * (root_k - last_root_k)) / (root_k * 5 + last_root_k).
+    /// `reserve0` - MUST be a positive number
+    /// `reserve1` - MUST be a positive number
+    /// Arguments:
+    /// * `reserve0` - the available amount of token0
+    /// * `reserve1` - the available amount of token1
+    async fn _mint_fee(&mut self, reserve0: u128, reserve1: u128) -> bool {
+        // get fee_to from factory
+        let fee_to: ActorId = messages::get_fee_to(&self.factory).await;
+        let fee_on = fee_to != ZERO_ID;
+        if fee_on {
+            if self.k_last != 0 {
+                let root_k = reserve0.overflowing_mul(reserve1).0.sqrt();
+                let root_k_last = self.k_last.sqrt();
+                if root_k > root_k_last {
+                    let numerator = self
+                        .get()
+                        .total_supply
+                        .overflowing_mul(root_k.overflowing_sub(root_k_last).0)
+                        .0;
+                    let denominator = root_k.overflowing_mul(5).0.overflowing_add(root_k_last).0;
+                    let liquidity = numerator.overflowing_div(denominator).0;
+                    if liquidity > 0 {
+                        FTCore::mint(self, &fee_to, liquidity);
+                    }
+                }
+            }
+        } else if self.k_last != 0 {
+            self.k_last = 0;
+        }
+        fee_on
+    }
+
+    /// Updates reserves and, on the first call per block, price accumulators
+    /// `balance0` - MUST be a positive number
+    /// `balance1` - MUST be a positive number
+    /// `reserve0` - MUST be a positive number
+    /// `reserve1` - MUST be a positive number
+    /// Arguments:
+    /// * `balance0` - token0 balance
+    /// * `balance1` - token1 balance
+    /// * `reserve0` - the available amount of token0
+    /// * `reserve1` - the available amount of token1
+    fn _update(&mut self, balance0: u128, balance1: u128, reserve0: u128, reserve1: u128) {
+        let current_ts = exec::block_timestamp() % (1 << 32);
```

Comment:
Too magic. What does it mean?

Reply:
Originally is just taking a reminder of the current timestamp where the modulo is 2^32.